flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java

```diff
@@ -0,0 +1,177 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.contrib.streaming.state;
+
+import org.apache.flink.core.fs.CloseableRegistry;
+import org.apache.flink.core.fs.FSDataInputStream;
+import org.apache.flink.core.fs.FSDataOutputStream;
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.runtime.concurrent.FutureUtils;
+import org.apache.flink.runtime.state.IncrementalKeyedStateHandle;
+import org.apache.flink.runtime.state.StateHandleID;
+import org.apache.flink.runtime.state.StreamStateHandle;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * Data transfer utils for RocksDbKeyedStateBackend.
+ */
+public class RocksDbStateDataTransfer {
+
+	public static void transferAllStateDataToDirectory(
+		IncrementalKeyedStateHandle restoreStateHandle,
+		Path dest,
+		int restoringThreadNum,
+		CloseableRegistry closeableRegistry) throws Exception {
+
+		final Map<StateHandleID, StreamStateHandle> sstFiles =
+			restoreStateHandle.getSharedState();
+		final Map<StateHandleID, StreamStateHandle> miscFiles =
+			restoreStateHandle.getPrivateState();
+
+		downloadDataForAllStateHandles(sstFiles, dest, restoringThreadNum, closeableRegistry);
+		downloadDataForAllStateHandles(miscFiles, dest, restoringThreadNum, closeableRegistry);
+	}
+
+	/**
+	 * Copies all the files from the given stream state handles to the given path, renaming the files w.r.t. their
+	 * {@link StateHandleID}.
+	 */
+	private static void downloadDataForAllStateHandles(
+		Map<StateHandleID, StreamStateHandle> stateHandleMap,
+		Path restoreInstancePath,
+		int restoringThreadNum,
+		CloseableRegistry closeableRegistry
+		) throws Exception {
+
+		List<Runnable> runnables = createDownloadRunnables(stateHandleMap, restoreInstancePath, closeableRegistry);
+
+		if (enableMultiThreadDownload(restoringThreadNum)) {
+			runTransferWithMultipleThreads(runnables, restoringThreadNum, closeableRegistry);
+		} else {
+			runTransferWithSingleThread(runnables);
+		}
```

Comment:
Wouldn't it be better to first instantiate the `ExecutorService` and then use it to download the files.
```
ExecutorService executorService = createExecutorService(restoringThreadNum);

for (Runnable runnable: createDownloadRunnables()) {
    executorService.execute(runnable);
}
```

if `restoringThreadNum == 1`, then we create a direct executor service. That way we would not need to duplicate the logic in `runTransferWithSingleThread`.

Reply:
`DirectExecutorService` would execute everything in the calling thread.

Reply:
I agree, we will have less code by just using the `DirectExecutorService` instead of looping Runnable's. Then code in `runTransferWithMultipleThreads` could go back to `downloadDataForAllStateHandles` and `runTransferWithSingleThread` would not be needed anymore.

Reply:
I agree. I found `DirectExecutorService` in test code(`flink-runtime/src/test/java/org/apache/runtime/util/DirectExecutorService`), Do I miss something? or should I move it to `flink-runtime/src/main/java/org/apache/runtime/util/DirectExecutorService` ?